File: scripts/framework-audit-v2.mjs
#!/usr/bin/env node
/**
 * Refined Content Framework Audit — only meaningful issues
 */

import { promises as fs } from 'fs';

async function main() {
  const dataJs = await fs.readFile('src/app/blog/data.js', 'utf-8');

  // Parse all posts using regex
  const slugs = [...dataJs.matchAll(/slug:\s*"([^"]+)"/g)].map(m => m[1]);
  const titles = [...dataJs.matchAll(/title:\s*"([^"]+)"/g)].map(m => m[1]);
  const dates = [...dataJs.matchAll(/date:\s*"([^"]+)"/g)].map(m => m[1]);
  // More robust excerpt extraction
  const excerptRegex = /excerpt:\s*(?:\n\s*)?"([^"]+)"/g;
  const excerpts = [...dataJs.matchAll(excerptRegex)].map(m => m[1]);
  const authorRegex = /author:\s*"([^"]+)"/g;
  const authors = [...dataJs.matchAll(authorRegex)].map(m => m[1]);
  const tagMatches = [...dataJs.matchAll(/tags:\s*\[([^\]]+)\]/g)];
  const tagsList = tagMatches.map(m => m[1].split(',').map(t => t.trim().replace(/^"|"$/g, '')).filter(t => t));
  const contentMatches = [...dataJs.matchAll(/content:\s*`([\s\S]*?)`\s*,?/g)];
  const contents = contentMatches.map(m => m[1]);

  // Build post objects
  const posts = [];
  const count = Math.max(slugs.length, titles.length, dates.length, contents.length);
  for (let i = 0; i < count; i++) {
    posts.push({
      slug: slugs[i] || '?',
      title: titles[i] || '',
      date: dates[i] || '',
      excerpt: excerpts[i] || '',
      tags: tagsList[i] || [],
      author: authors[i] || '',
      content: contents[i] || '',
    });
  }

  // Clean content — trim leading non-content JS
  for (const p of posts) {
    let c = p.content;
    // Find first markdown heading or paragraph
    const firstContent = c.search(/^[#A-Za-z\u0980-\u09FF]/m);
    if (firstContent > 0 && firstContent < 200) c = c.substring(firstContent);
    // Remove trailing JS
    const lastContent = c.lastIndexOf('`');
    if (lastContent > c.length * 0.8) c = c.substring(0, lastContent);
    p.content = c.trim();
  }

  // ── REAL framework checks ──

  let reportSections = [];

  for (const post of posts) {
    if (!post.content || post.content.length < 50) continue;

    const c = post.content;
    const isBangla = /[\u0980-\u09FF]/.test(post.title);

    let issues = [];
    let checks = {};

    // A. TF-IDF — check keyword exists enough
    const stopwords = new Set(['the','a','an','in','for','of','to','and','is','are','what','how','why','when','where','your','our','its','their','complete','guide','ultimate','best','top','seo','vs','for','and','the']);
    const titleWords = post.title.replace(/[^a-zA-Z\s\u0980-\u09FF-]/g, '').split(/\s+/).filter(w => w.length > 2 && !stopwords.has(w.toLowerCase()));
    const keyword = titleWords.slice(0, 2).join(' ') || post.title.split(' ').slice(0, 2).join(' ');
    let kwCount = 0;
    if (keyword.length >= 3 && !isBangla) {
      const kwParts = keyword.toLowerCase().split(' ');
      for (const part of kwParts) {
        if (part.length > 3) {
          const re = new RegExp(part.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'gi');
          kwCount += (c.match(re) || []).length;
        }
      }
    } else if (isBangla) {
      // For Bangla posts, just check the main topic word appears
      // Extract first non-stopword from title
      const banglaWords = post.title.match(/[\u0980-\u09FF]+/g) || [];
      if (banglaWords.length > 0) {
        const mainWord = banglaWords[0];
        const re = new RegExp(mainWord, 'g');
        kwCount = (c.match(re) || []).length;
      } else {
        kwCount = 999; // skip
      }
    }
    const tfidfOk = kwCount >= 5 || isBangla;
    if (!tfidfOk && kwCount < 5 && kwCount > 0 && !isBangla) {
      issues.push(`🔑 TF-IDF: "${keyword}" only ${kwCount} occurrences (need 5+)`);
    }

    // B. Entity check — only meaningful ones
    let missingEntities = [];
    const lower = c.toLowerCase();
    const titleLower = post.title.toLowerCase();

    // Check location: if title mentions Dhaka/Bangladesh, content should too
    if (/dhaka/i.test(post.title) && !/dhaka/i.test(c)) missingEntities.push('Dhaka (in content)');
    if (/bangladesh/i.test(post.title) && !/bangladesh/i.test(c)) missingEntities.push('Bangladesh (in content)');

    // Author name
    if (!isBangla && !/kanok\s*miah/i.test(c) && !/kanok/i.test(c)) {
      // Check if it's purely a case study (UK-focused etc)
      if (!/locksmith|dundee|scotland|uk|landlord/i.test(post.slug)) {
        missingEntities.push('Kanok Miah (author branding)');
      }
    }

    if (missingEntities.length > 0) {
      issues.push(`🏷️ Missing entities: ${missingEntities.join(', ')}`);
    }

    // C. Pillar link check
    const tags = (post.tags || []).join(' ').toLowerCase();
    let pillarSlug = null;
    let pillarName = null;

    if (tags.includes('seo guide') || tags.includes('bangladesh seo')) {
      pillarSlug = 'complete-seo-guide-bangladesh-businesses-2026';
      pillarName = 'Complete SEO Guide';
    } else if (tags.includes('ecommerce') || tags.includes('e-commerce')) {
      pillarSlug = 'why-ecommerce-store-needs-seo-bangladesh';
      pillarName = 'E-commerce SEO guide';
    } else if (tags.includes('technical')) {
      pillarSlug = 'technical-seo-checklist-bangladeshi-websites';
      pillarName = 'Technical SEO guide';
    } else if (tags.includes('local')) {
      pillarSlug = 'google-business-profile-optimization-guide-bangladesh';
      pillarName = 'GBP optimization guide';
    } else if (tags.includes('link building')) {
      pillarSlug = 'link-building-strategies-bangladesh-market';
      pillarName = 'Link Building guide';
    } else if (tags.includes('mobile')) {
      pillarSlug = 'mobile-seo-optimization-bangladesh-mobile-first-era';
      pillarName = 'Mobile SEO guide';
    } else if (tags.includes('content')) {
      pillarSlug = 'content-marketing-strategy-bangladeshi-brands-seo';
      pillarName = 'Content Marketing guide';
    } else if (tags.includes('geo') || tags.includes('aeo') || tags.includes('ai search') || tags.includes('generative engine')) {
      pillarSlug = 'geo-optimization-prepare-business-ai-search';
      pillarName = 'GEO/AEO guide';
    } else if (tags.includes('garment') || tags.includes('textile')) {
      pillarSlug = 'seo-garments-textile-industry-b2b-lead-generation';
      pillarName = 'Garments/Textile SEO guide';
    } else if (tags.includes('real estate')) {
      pillarSlug = 'seo-real-estate-developers-dhaka';
      pillarName = 'Real Estate SEO guide';
    } else if (tags.includes('international') || tags.includes('export')) {
      pillarSlug = 'international-seo-bangladesh-exporters-global-buyers';
      pillarName = 'International SEO guide';
    } else if (tags.includes('google ads') || tags.includes('ppc')) {
      pillarSlug = 'seo-vs-google-ads-bangladesh-business';
      pillarName = 'SEO vs Google Ads guide';
    } else if (tags.includes('schema') || tags.includes('structured data') || tags.includes('rich snippet')) {
      pillarSlug = 'schema-markup-rich-snippets-techniques';
      pillarName = 'Schema markup guide';
    } else if (tags.includes('youtube') || tags.includes('video')) {
      pillarSlug = 'youtube-seo-bangladesh-ranking-tips';
      pillarName = 'YouTube SEO guide';
    } else if (tags.includes('bangla') || tags.includes('বাংলা') || tags.includes('বাংলাদেশ')) {
      pillarSlug = 'seo-bangla-beginners-guide-google-ranking';
      pillarName = 'SEO Bangla guide';
    } else if (tags.includes('case study')) {
      pillarSlug = null; // case studies have flexible pillar mapping
    } else if (tags.includes('keyword research')) {
      pillarSlug = 'complete-seo-guide-bangladesh-businesses-2026';
      pillarName = 'Complete SEO Guide';
    } else if (tags.includes('voice search')) {
      pillarSlug = 'mobile-seo-optimization-bangladesh-mobile-first-era';
      pillarName = 'Mobile SEO guide';
    } else if (tags.includes('google business profile') || tags.includes('gbp') || tags.includes('google my business')) {
      pillarSlug = 'google-business-profile-optimization-guide-bangladesh';
      pillarName = 'GBP optimization guide';
    } else if (tags.includes('web development') || tags.includes('web design')) {
      pillarSlug = 'seo-for-new-website-bangladesh';
      pillarName = 'New Website SEO guide';
    } else if (tags.includes('blogging')) {
      pillarSlug = 'content-marketing-strategy-bangladeshi-brands-seo';
      pillarName = 'Content Marketing guide';
    }

    if (pillarSlug) {
      if (post.slug !== pillarSlug) {
        const pillarLinked = c.includes(pillarSlug);
        if (!pillarLinked) {
          issues.push(`🔗 Missing pillar link: link to "${pillarName}" post (/blog/${pillarSlug})`);
        }
      }
    } else {
      // For posts with no clear pillar mapping (case studies, industry-specific), skip silently
    }

    // D. AEO/GEO — question headings (only for English posts, Bangla posts naturally don't use English question H2s)
    if (!isBangla) {
      const questionH2 = (c.match(/^#{1,4}\s+(How|What|Why|When|Where|Can|Do|Is|Are|Does|Which|Who)\b/gim) || []);
      if (questionH2.length < 2) {
        issues.push(`❓ AEO/GEO: Only ${questionH2.length} question-based headings (need 2+ for AI search snippets)`);
      }
    }

    // E. Internal links
    const internalLinks = (c.match(/\]\(\/(?!http|https|www\.)(?:blog\/|services\/|locations\/|industries\/|about|contact)/g) || []);
    // Also count links to / (homepage)
    const homeLinks = (c.match(/\]\(\//g) || []);
    // Filter: links that are just / (homepage) vs full paths
    const homeOnlyLinks = (c.match(/\]\(\/\)/g) || []).length;
    const totalRelevantLinks = internalLinks.length + homeOnlyLinks;
    
    // For case studies/about pages, 2+ links is acceptable if they are relevant
    const isCaseStudy = post.tags.some(t => t.toLowerCase().includes('case study')) || post.slug.includes('case-study');
    const minLinks = isCaseStudy ? 2 : 3;
    
    // Exclude homepage links from the count since they were just added by the linking audit
    const nonHomeLinks = internalLinks.length;
    
    if (nonHomeLinks < minLinks && !isCaseStudy) {
      issues.push(`🔗 Internal links: only ${nonHomeLinks} to blog/services/locations (need ${minLinks}+)`);
    } else if (nonHomeLinks < 2 && isCaseStudy) {
      issues.push(`🔗 Internal links: only ${nonHomeLinks} to blog/services/locations (need ${minLinks}+)`);
    }

    // F. Schema readiness
    let schemaMissing = [];
    if (!post.title || post.title.trim() === '') schemaMissing.push('title');
    if (!post.date || post.date.trim() === '') schemaMissing.push('date');
    if (!post.author || post.author.trim() === '') schemaMissing.push('author');
    if (!post.slug || post.slug.trim() === '') schemaMissing.push('slug');
    if (schemaMissing.length > 0) {
      issues.push(`📋 Schema: missing field(s): ${schemaMissing.join(', ')}`);
    }

    // Only output if there are issues
    if (issues.length > 0) {
      reportSections.push({ slug: post.slug, title: post.title, issues, isBangla });
    }
  }

  // ── Generate report ──

  console.log('# Content Framework Enforcement Report');
  console.log(`Generated: ${new Date().toISOString()}`);
  console.log(`Posts audited: ${posts.length}`);
  console.log(`Posts with issues: ${reportSections.length}`);
  console.log(`Posts clean: ${posts.length - reportSections.length}`);
  console.log('');

  // Group by issue type for summary
  const typeCounts = {};
  for (const sec of reportSections) {
    for (const issue of sec.issues) {
      const type = issue.split(':')[0];
      typeCounts[type] = (typeCounts[type] || 0) + 1;
    }
  }

  console.log('## Summary by Issue Type');
  const sortedTypes = Object.entries(typeCounts).sort((a, b) => b[1] - a[1]);
  for (const [type, count] of sortedTypes) {
    console.log(`- ${type}: ${count} posts affected`);
  }
  console.log('');

  // Individual post reports
  console.log('## Per-Post Issues');
  console.log('');

  for (const sec of reportSections) {
    console.log(`### ${sec.slug}`);
    if (sec.title.length < 80) console.log(`*${sec.title}*`);
    console.log('');
    for (const issue of sec.issues) {
      console.log(`- ${issue}`);
    }
    console.log('');
  }

  // Priority action items
  console.log('## ⚡ Priority Fixes (by impact)');
  console.log('');

  // Count missing excerpts
  const missingExcerpt = reportSections.filter(s => s.issues.some(i => i.includes('Schema: missing')));
  const missingPillar = reportSections.filter(s => s.issues.some(i => i.includes('Missing pillar link')));
  const missingAEO = reportSections.filter(s => s.issues.some(i => i.includes('AEO/GEO')));
  const missingLinks = reportSections.filter(s => s.issues.some(i => i.includes('Internal links:')));
  const missingEntitiesReport = reportSections.filter(s => s.issues.some(i => i.includes('Missing entities')));

  if (missingExcerpt.length > 0) {
    console.log(`1️⃣  Add missing excerpt fields (${missingExcerpt.length} posts)`);
    console.log(`   Posts: ${missingExcerpt.map(s => s.slug).join(', ')}`);
    console.log('   → ArticleSchema requires excerpt. Open each post in data.js and add:\n     excerpt: "Your SEO-optimized meta description here.",');
    console.log('');
  }

  if (missingPillar.length > 0) {
    console.log(`2️⃣  Add pillar links (${missingPillar.length} posts)`);
    console.log(`   Posts: ${missingPillar.map(s => s.slug).join(', ')}`);
    console.log('   → Each cluster post should link back to its pillar page.');
    console.log('');
  }

  if (missingAEO.length > 0) {
    console.log(`3️⃣  Add question-based H2 headings for AEO/GEO (${missingAEO.length} English posts)`);
    console.log(`   Posts: ${missingAEO.map(s => s.slug).join(', ')}`);
    console.log('   → AI search engines favor content with clear Q&A sections.');
    console.log('   → Add 2+ H2s starting with How, What, Why, When, Where, Can, Do, Is, Are');
    console.log('');
  }

  if (missingLinks.length > 0) {
    console.log(`4️⃣  Increase internal linking (${missingLinks.length} posts)`);
    console.log(`   Posts: ${missingLinks.map(s => s.slug).join(', ')}`);
    console.log('   → Add links to /blog/, /services/, /locations/ within content body.');
    console.log('');
  }

  if (missingEntitiesReport.length > 0) {
    console.log(`5️⃣  Entity coverage improvements (${missingEntitiesReport.length} posts)`);
    console.log('   → Ensure location (Dhaka/Bangladesh) and author brand (Kanok Miah) are mentioned.');
    console.log('');
  }

  console.log('---');
  console.log(`✅ ${posts.length - reportSections.length} posts passed all checks.`);
}

main().catch(e => {
  console.error('Error:', e);
  process.exit(1);
});

File: framework_audit_v2.py
#!/usr/bin/env python3
"""
Content Framework Enforcer for kanokmiah.com.bd
Line-based parser for large JS data file.
"""
import re
import sys

with open('/root/kanok-miahit/src/app/blog/data.js', 'r') as f:
    lines = f.readlines()

print(f"Total lines: {len(lines)}", file=sys.stderr)

# Parse posts using line-by-line state machine
posts = []
current_post = None
in_content = False
content_lines = []

for i, line in enumerate(lines):
    stripped = line.strip()
    
    # Detect slug line
    if not in_content and stripped.startswith('slug:'):
        slug = stripped.split('"')[1] if '"' in stripped else ''
        current_post = {'slug': slug, 'title': '', 'date': '', 'excerpt': '', 'tags': [], 'content': '', 'line_start': i + 1}
    
    # Detect title
    if current_post and not in_content and stripped.startswith('title:'):
        title_match = re.match(r'title:\s*"(.+)"\s*,?\s*', stripped)
        if title_match:
            current_post['title'] = title_match.group(1)
    
    # Detect date
    if current_post and not in_content and stripped.startswith('date:'):
        date_match = re.match(r'date:\s*"(.+)"\s*,?\s*', stripped)
        if date_match:
            current_post['date'] = date_match.group(1)
    
    # Detect excerpt
    if current_post and not in_content and stripped.startswith('excerpt:'):
        # May span multiple lines
        excerpt = stripped.replace('excerpt:', '').strip().strip(',').strip()
        if excerpt.startswith('"') and excerpt.endswith('"'):
            current_post['excerpt'] = excerpt.strip('"')
        else:
            # Multi-line excerpt - collect
            excerpt_lines = [excerpt]
            j = i + 1
            while j < len(lines):
                l = lines[j].strip().rstrip(',')
                excerpt_lines.append(l)
                if l.endswith('"'):
                    break
                j += 1
            full = ' '.join(excerpt_lines).strip().strip('"')
            current_post['excerpt'] = full
    
    # Detect tags
    if current_post and not in_content and stripped.startswith('tags:'):
        tags_str = stripped.replace('tags:', '').strip().strip(',').strip()
        if tags_str.startswith('[') and tags_str.endswith(']'):
            # Single line
            tags = re.findall(r'"([^"]+)"', tags_str)
            current_post['tags'] = tags
        else:
            # Multi-line
            tags_parts = [tags_str]
            j = i + 1
            while j < len(lines):
                l = lines[j].strip().rstrip(',')
                tags_parts.append(l)
                if l.endswith(']'):
                    break
                j += 1
            full = ' '.join(tags_parts)
            tags = re.findall(r'"([^"]+)"', full)
            current_post['tags'] = tags
    
    # Detect content start
    if current_post and not in_content and 'content:' in stripped and stripped.endswith('`') and not stripped.startswith('#'):
        in_content = True
        content_lines = []
        # Check if content is on same line as ` or on next line
        if stripped.strip() == 'content: `' or stripped.strip() == 'content:`':
            # Content starts on next line
            pass
        else:
            # Content starts inline after `content: ` - extract text after backtick
            content_start = stripped.find('`') + 1
            rest = stripped[content_start:]
            if rest:
                # Remove trailing comma and comment
                rest = re.sub(r'`.*$', '', rest)
                content_lines.append(rest)
    
    # Collect content lines
    if in_content:
        # Check if this line (or its stripped version) ends the content
        # Look for closing backtick followed by comma (possibly with comment)
        end_match = re.match(r'^(.*?)`\s*,?\s*(//.*)?$', stripped)
        if end_match and end_match.group(1) is not None:
            # Might be the end
            before_backtick = end_match.group(1)
            if before_backtick or len(content_lines) > 0:
                # End of content
                if before_backtick:
                    content_lines.append(before_backtick)
                current_post['content'] = '\n'.join(content_lines)
                current_post['line_end'] = i + 1
                posts.append(current_post)
                current_post = None
                in_content = False
                content_lines = []
                continue
        else:
            # Check if line is just closing backtick
            if stripped.startswith('`'):
                # This line starts with backtick - it's the closing delimiter
                # Content is everything collected so far
                content_after = stripped[1:].strip()
                if content_after and not content_after.startswith(','):
                    # There's text after the backtick - add it?
                    # Actually this shouldn't happen in template literals
                    pass
                current_post['content'] = '\n'.join(content_lines)
                current_post['line_end'] = i + 1
                posts.append(current_post)
                current_post = None
                in_content = False
                content_lines = []
                continue
            else:
                content_lines.append(stripped)

print(f"Parsed posts: {len(posts)}", file=sys.stderr)

# Verify
for p in posts[:3]:
    print(f"  {p['slug']}: {p['title'][:50]}... content length={len(p['content'])}", file=sys.stderr)

# Check for which slugs were modified
modified_slugs = {
    # From commit 001ef98 (internal linking)
    "seo-people-also-ask-optimization",
    "seo-featured-snippet-bangladesh", 
    "seo-knowledge-panel-bangladesh",
    "locksmith-dundee-seo-case-study",
    "das-taxis-scotland-seo-case-study",
    "morethanpanel-seo-case-study",
    "smmgen-seo-case-study",
    "smmsun-seo-case-study",
    "mir-cement-seo-case-study",
    "dhaka-apparels-seo-case-study",
    "stealth-windshield-repairs-seo-case-study",
    "seo-expert-vs-seo-agency-dhaka-which-is-right",
    "top-10-seo-mistakes-dhaka-businesses-fix",
    "seo-case-study-dhaka-businesses-increased-organic-traffic",
    "hiring-seo-expert-dhaka-better-roi-than-paid-ads",
    "watchzonebd-seo-case-study",
    # From commit cad9c06 (blank line removal) - extract from diff
    "complete-seo-guide-bangladesh-businesses-2026",
    "why-ecommerce-store-needs-seo-bangladesh",
    "technical-seo-checklist-bangladeshi-websites",
    "how-to-choose-right-seo-agency-bangladesh",
    "link-building-strategies-bangladesh-market",
    "geo-optimization-prepare-business-ai-search",
    "seo-garments-textile-industry-b2b-lead-generation",
    "google-business-profile-optimization-guide-bangladesh",
    "seo-vs-google-ads-whats-best-bangladesh-businesses",
    "seo-real-estate-developers-dhaka",
    "mobile-seo-optimization-bangladesh-mobile-first-era",
    "content-marketing-strategy-bangladeshi-brands-seo",
    "international-seo-bangladesh-exporters-global-buyers",
    "seo-bangla-beginners-guide-google-ranking",
    "local-seo-dhaka-google-maps-ranking",
    # From commit 5cbb3f7 (HTML entity fix)
    "schema-markup-rich-snippets-techniques",
    "seo-canonical-url-guide-bd",
}

# Build lookup
posts_by_slug = {p['slug']: p for p in posts}

# Check which modified slugs are found
found_slugs = []
not_found_slugs = []
for s in sorted(modified_slugs):
    if s in posts_by_slug:
        found_slugs.append(s)
    else:
        not_found_slugs.append(s)

if not_found_slugs:
    print(f"NOT FOUND in parsed data: {not_found_slugs}", file=sys.stderr)

print(f"Found slugs to check: {found_slugs}", file=sys.stderr)

# =============================================
# Framework Checks
# =============================================

def extract_primary_keyword(title):
    """Extract primary keyword from title"""
    t = title.lower()
    # Remove common prefixes
    for prefix in ['complete ', 'why your ', 'how to ', 'what is ', 'top ', 'best ', 
                   'the ', 'a ', 'an ', 'ultimate ', 'seo ']:
        if t.startswith(prefix):
            t = t[len(prefix):]
            break
    # Remove trailing location/context qualifiers
    t = re.sub(r'\s+(in|for|of|—|:).*$', '', t)
    # Take first 2-3 meaningful words
    words = [w for w in t.split() if len(w) > 2][:3]
    if not words:
        words = t.split()[:3]
    return ' '.join(words) if words else title.split()[:3]

def count_occurrences(text, keyword):
    if not keyword:
        return 0
    return len(re.findall(re.escape(keyword), text, re.IGNORECASE))

def check_entities(content):
    """Check semantic entity coverage"""
    location_found = {
        'Dhaka': bool(re.search(r'\bDhaka\b', content)),
        'Bangladesh': bool(re.search(r'\bBangladesh\b', content)),
    }
    service_terms = ['SEO', 'search engine optimization', 'digital marketing', 
                     'local SEO', 'technical SEO', 'link building', 'content marketing',
                     'on-page', 'off-page', 'Google Business']
    service_found = any(re.search(term, content, re.IGNORECASE) for term in service_terms)
    
    industry_terms = ['e-commerce', 'ecommerce', 'real estate', 'healthcare', 'medical',
                      'garment', 'textile', 'education', 'restaurant', 'hotel', 'travel',
                      'B2B', 'startup', 'legal', 'fitness', 'photography', 'event',
                      'non-profit', 'NGO', 'export', 'cleaning', 'salon']
    industry_found = any(re.search(term, content, re.IGNORECASE) for term in industry_terms)
    
    missing = []
    if not location_found['Dhaka']:
        missing.append('Dhaka')
    if not location_found['Bangladesh']:
        missing.append('Bangladesh')
    if not service_found:
        missing.append('service_type')
    
    return missing, {'Dhaka': location_found['Dhaka'], 'Bangladesh': location_found['Bangladesh'], 'service': service_found, 'industry': industry_found}

def find_pillar_links(content, tags, slug):
    """Check pillar page links"""
    pillar_map = {
        'SEO Guide': 'complete-seo-guide-bangladesh-businesses-2026',
        'Local SEO': 'local-seo-tips-dhaka-businesses-google-maps',
        'Technical SEO': 'technical-seo-checklist-bangladeshi-websites',
        'E-commerce SEO': 'why-ecommerce-store-needs-seo-bangladesh',
        'Link Building': 'link-building-strategies-bangladesh-market',
        'GEO': 'geo-optimization-prepare-business-ai-search',
        'Mobile SEO': 'mobile-seo-optimization-bangladesh-mobile-first-era',
        'Content Marketing': 'content-marketing-strategy-bangladeshi-brands-seo',
        'International SEO': 'international-seo-bangladesh-exporters-global-buyers',
        'Google Ads': 'seo-vs-google-ads-whats-best-bangladesh-businesses',
        'Real Estate': 'seo-real-estate-developers-dhaka',
        'Google Business': 'google-business-profile-optimization-guide-bangladesh',
        'Garments SEO': 'seo-garments-textile-industry-b2b-lead-generation',
        'Bangladesh SEO': 'complete-seo-guide-bangladesh-businesses-2026',
        'Digital Marketing': 'complete-seo-guide-bangladesh-businesses-2026',
        '2026': 'seo-trends-2026-ai-geo-future',
    }
    
    linked = []
    for tag in tags:
        if tag in pillar_map:
            pillar_slug = pillar_map[tag]
            pillar_path = f'/blog/{pillar_slug}'
            if pillar_slug != slug and pillar_path in content:
                linked.append(pillar_path)
    
    # Also check for homepage link (/)
    if '(/)' in content or '(/)' in content:
        linked.append('/ (homepage)')
    
    return linked

def count_question_headings(content):
    q_words = r'(?:How|What|Why|When|Where|Can|Do|Is|Are|Does|Which|Who|Should)'
    headings = re.findall(r'^#{1,6}\s+' + q_words + r'\b', content, re.MULTILINE)
    return len(headings)

def count_internal_links(content, slug):
    links = re.findall(r'\((/(?:blog|services|locations|industries|about|contact)/?[^)]*)\)', content)
    # Filter out self-references
    non_self = [l for l in links if f'/blog/{slug}' not in l]
    unique = set(non_self)
    return len(unique)

def check_schema_readiness(post):
    issues = []
    if not post.get('title', '').strip():
        issues.append('title')
    excerpt = post.get('excerpt', '').strip()
    if not excerpt:
        issues.append('excerpt')
    if not post.get('date', '').strip():
        issues.append('date')
    return issues

# Generate report
print("# Content Framework Audit Report")
print(f"**Generated:** Cron job — 2026-07-21")
print(f"**Trigger:** 3 commits in last 48 hours with changes to blog posts")
print(f"**Posts analyzed:** {len(found_slugs)}")
print()

all_pass = True

for slug in sorted(found_slugs):
    post = posts_by_slug[slug]
    content = post['content']
    title = post['title']
    tags = post['tags']
    
    print(f"## Post: {slug}")
    print(f"**Title:** {title}")
    print(f"**Tags:** {', '.join(tags)}")
    print()
    print("| Check | Status | Details |")
    print("|-------|--------|---------|")
    
    # A. TF-IDF Coverage
    keyword = extract_primary_keyword(title)
    kw_count = count_occurrences(content, keyword)
    kw_status = "✅" if kw_count >= 5 else "❌"
    if kw_status == "❌":
        all_pass = False
    print(f"| TF-IDF: \"{keyword}\" | {kw_status} | {kw_count} occurrences |")
    
    # B. Entities
    missing_entities, entity_details = check_entities(content)
    entity_status = "✅" if len(missing_entities) == 0 else "❌"
    if entity_status == "❌":
        all_pass = False
    entity_detail = "All key entities present" if not missing_entities else f"Missing: {', '.join(missing_entities)}"
    print(f"| Entities | {entity_status} | {entity_detail} |")
    
    # C. Pillar Link
    pillar_links = find_pillar_links(content, tags, slug)
    pillar_status = "✅" if pillar_links else "❌"
    if pillar_status == "❌":
        all_pass = False
    pillar_detail = ', '.join(pillar_links) if pillar_links else "No pillar link found"
    print(f"| Pillar Link | {pillar_status} | {pillar_detail} |")
    
    # D. AEO/GEO
    q_count = count_question_headings(content)
    q_status = "✅" if q_count >= 2 else "❌"
    if q_status == "❌":
        all_pass = False
    print(f"| AEO/GEO | {q_status} | {q_count} question headings |")
    
    # E. Internal Links
    link_count = count_internal_links(content, slug)
    link_status = "✅" if link_count >= 3 else "❌"
    if link_status == "❌":
        all_pass = False
    print(f"| Internal Links | {link_status} | {link_count} unique internal links |")
    
    # F. Schema
    schema_issues = check_schema_readiness(post)
    schema_status = "✅" if len(schema_issues) == 0 else "❌"
    if schema_status == "❌":
        all_pass = False
    schema_detail = "All fields set" if not schema_issues else f"Missing: {', '.join(schema_issues)}"
    print(f"| Schema Ready | {schema_status} | {schema_detail} |")
    
    print()

print("---")
if all_pass:
    print("## ✅ Overall: All framework checks pass!")
else:
    print("## ⚠ Overall: Some posts need attention (flagged above)")
print()
print("### Legend")
print("- **TF-IDF:** Primary keyword should appear ≥5 times for topical relevance")
print("- **Entities:** Location (Dhaka & Bangladesh), service type, industry context must be present")
print("- **Pillar Link:** Posts should link back to their pillar/topic cluster hub")
print("- **AEO/GEO:** ≥2 question-based headings for AI/voice search optimization")
print("- **Internal Links:** ≥3 internal links to other posts, services, or locations")
print("- **Schema Ready:** title + excerpt + date needed for ArticleSchema markup")
